src/algorithm/REDUCE_SUM-Cuda.cpp

```diff
@@ -158,14 +180,21 @@ void REDUCE_SUM::runCudaVariantBlock(VariantID vid)
                                                    iend );
       cudaErrchk( cudaGetLastError() );
 
-      cudaErrchk( cudaMemcpyAsync( &m_sum, dsum, sizeof(Real_type),
-                                   cudaMemcpyDeviceToHost, res.get_stream() ) );
+      if (separate_buffers) {
+        cudaErrchk( cudaMemcpyAsync( hsum, dsum, sizeof(Real_type),
+                                     cudaMemcpyDeviceToHost, res.get_stream() ) );
+      }
+
       cudaErrchk( cudaStreamSynchronize( res.get_stream() ) );
+      m_sum = *hsum;
 
     }
     stopTimer();
 
-    deallocData(DataSpace::CudaDevice, dsum);
+    deallocData(rds, dsum);
+    if (separate_buffers) {
+      deallocData(hrds, hsum);
+    }
```

Comment:
...to this line are representative of the changes I am proposing.

Reply:
Can we make these changes cleaner in a macro or function to avoid a lot of new redundant code?

Reply:
Ya we could do that for some of this, at least the setup and tear-down at the top and bottom.

Reply:
I put the GPU reducer setup and teardown in a macro. That cleans up these sections of code for the reducers but PI_ATOMIC breaks symmetry as it allocates its buffer slightly differently.

Reply:
The memory copies are still copied, but they do vary by kernel and backend a bit more. Do you think its worth abstracting them too?